### PR TITLE
refactor (app): Bitcoin transaction porting

### PR DESCRIPTION
**Pull request by ujjwal-cyph**
The changeset adds and connects the transaction signing flow for the Bitcoin family. Only the initial host queries are included in this PR. A follow-up change (#332) will add rest of the query handling.

Implements https://app.clickup.com/t/9002019994/PRF-4814

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-4814 Send transaction support](https://app.clickup.com/t/9002019994/PRF-4814)

File: apps/btc_family/btc_txn.c [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  memcpy(&g_txn_context->init_info,
+         &query->sign_txn.initiate,
+         sizeof(btc_sign_txn_initiate_request_t));
+  send_response(BTC_SIGN_TXN_RESPONSE_CONFIRMATION_TAG);
+  return true;
+}
+
+static bool fetch_transaction_meta(btc_query_t *query) {
+  if (!btc_get_query(query, BTC_QUERY_SIGN_TXN_TAG) ||
+      !check_which_request(query, BTC_SIGN_TXN_REQUEST_META_TAG)) {
```
Comment (kartikCypherock) [inline apps/btc_family/btc_txn.c:247]: Any check required to validate the meta data? 

I am guessing we will put a limit on # of inputs and outputs.. so ideally it should go here

File: apps/btc_family/btc_txn.c [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+  }
+
+  if (NULL != g_txn_context && NULL != g_txn_context->inputs) {
+    free(g_txn_context->inputs);
+  }
+  if (NULL != g_txn_context && NULL != g_txn_context->outputs) {
+    free(g_txn_context->outputs);
+  }
+  if (NULL != g_txn_context) {
+    free(g_txn_context);
```
Comment (kartikCypherock) [inline apps/btc_family/btc_txn.c:284]:
```suggestion
    free(g_txn_context);
    g_txn_context = NULL;
```

File: apps/btc_family/btc_txn.c [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+ * @retval false If the query does not contain the expected request
+ */
+static bool check_which_request(const btc_query_t *query,
+                                pb_size_t which_request);
+
+/**
+ * @brief Validates all the derivation paths received in the request from host
+ * @details The function validates each path entry in the request. If any
+ * invalid path is detected, the function will send an error to the host and
+ * return false.
```
Comment (kartikCypherock) [inline apps/btc_family/btc_txn.c:128]:
```suggestion
 * @brief Validates the derivation path received in the request from host. This derivation path will be used during the signing process.
 * @details The function validates path entry in the request. If any
 * invalid path is detected, the function will send an error to the host and
 * return false.
```

File: apps/btc_family/btc_txn.c (hunk 1 above)
Comment (ujjwal-cyph) [reply apps/btc_family/btc_txn.c:247]: this is added in #332. can add a todo here

**Review by kartikCypherock — APPROVED** [review]

**Review by amanCypherock — APPROVED** [review]

File: apps/btc_family/btc_txn.c [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+ *
+ * @return bool Indicating if the function actions succeeded or failed
+ */
+static bool fetch_transaction_meta(btc_query_t *query);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+static txn_context *g_txn_context = NULL;
```
Comment (amanCypherock) [inline apps/btc_family/btc_txn.c:179]: Consider renaming to `btc_txn_context`.
`g_` comes from Hungarian notation and is disallowed in coding guidelines.

Comment (ujjwal-cyph) [reply apps/btc_family/btc_txn.c]: changed. thanks

Comment (ujjwal-cyph) [reply apps/btc_family/btc_txn.c]: resolving this and merging.

**Pull request merged**